torch/_inductor/simple_fsdp/utils.py

```diff
@@ -0,0 +1,114 @@
+import math
+from enum import IntEnum
+from collections import defaultdict
+from typing import Dict, List, Optional, Set, Tuple
+
+import torch
+import torch.distributed as dist
+from .. import dependencies, ir, scheduler
+from ..virtualized import V
+
+class NodeType(IntEnum):
+    ALL_GATHER = 0
+    COMPUTE = 1
+    REDUCE_SCATTER = 2
+    AG_WAIT = 3
+    RS_WAIT = 4
+
+def compute_bucket_users(
```

Comment:
didn't understand the code upon first glimpse, let's go over the details together some time.